File: 2017/cpp/day_02a.cpp
#include <fstream>
#include <iostream>
#include <string>
#include <charconv>
#include <limits>
#include <ranges>

int main(int argc, char* argv[]) {
  const std::string input = (argc > 1) ? argv[1] : "../input/day_02_input" ;  
  std::ifstream file(input);
  std::string line;
  int checksum = 0; 
  while(std::getline(file, line)) {
    auto numbers = line 
                    | std::ranges::views::split(' ') 
                    | std::ranges::views::transform([](auto&& rng) { 
                        return std::string_view(&*rng.begin(), std::ranges::distance(rng)); });
    int min_val = std::numeric_limits<int>::max();
    int max_val = std::numeric_limits<int>::min();
    for (const auto n : numbers) {
        int val;
        std::from_chars(n.data(), n.data() + n.size(), val);
        max_val = std::max(val, max_val);
        min_val = std::min(val, min_val);
    }
    checksum += max_val - min_val;
  }
  std::cout << checksum << '\n';
  return 0;
}

File: 2017/py/day02a.py
import sys

def calculate_checksum(filename):
      checksum = 0
      with open(filename, 'r') as file:
          for line in file:
              # 使用split方法处理可能存在的制表符或其他分隔符
              numbers = line.strip().split()
              if not numbers:
                  continue # 跳过空行
              min_val = max_val = int(numbers[0])
              for number in numbers[1:]:
                  val = int(number)
                  max_val = max(val, max_val)
                  min_val = min(val, min_val)
              checksum += max_val - min_val
      return checksum

if __name__ == "__main__":
      input_file = sys.argv[1] if len(sys.argv) > 1 else "input"
      checksum = calculate_checksum(input_file)
      print(checksum)
